### PR TITLE
feat: add negative compare-and-delete to imemcache

### PR DESCRIPTION
**Pull request by icewind1991**
Delete the key if the value does not equal some value.

Extracted from sharding work

### CODE REVIEW

**Review by sorbaugh — APPROVED** [review]

File: lib/public/IMemcache.php [hunk 1]
```diff
@@ -67,10 +67,24 @@ public function cas($key, $old, $new);
...
 	public function cad($key, $old);
+
+	/**
+	 * Negative compare and delete
+	 *
+	 * Delete $key if the stored value is not equal to $old
+	 *
+	 * @param string $key
+	 * @param mixed $old
+	 * @return bool
```
Comment (come-nc) [inline lib/public/IMemcache.php:86]: Add a description for the return, what does the bool mean? false is failure or not found?

File: lib/public/IMemcache.php [hunk 2]
```diff
@@ -67,10 +67,24 @@ public function cas($key, $old, $new);
...
+	 * @since 30.0.0
+	 */
+	public function ncad(string $key, $old): bool;
```
Comment (come-nc) [inline lib/public/IMemcache.php:89]:
```suggestion
	public function ncad(string $key, mixed $old): bool;
```
In these cases where we really accept anything we should strongtype it as well, `mixed` is available in `PHP>=8.0`

File: lib/private/Memcache/CADTrait.php [hunk 3]
```diff
@@ -35,4 +35,20 @@ public function cad($key, $old) {
 			return false;
 		}
 	}
+
+	public function ncad(string $key, $old): bool {
+		//no native cas, emulate with locking
```
Comment (come-nc) [inline lib/private/Memcache/CADTrait.php:40]: What is cas, is that a typo for cad?

**Review by come-nc — APPROVED** [review]

**Pull request merged**